### ISSUE TITLE
An unexpected error during async rewrite of generic compound assignment to a member of an RValue

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#79415]
```
using System.Threading.Tasks;

interface I1 
{
    int this[int i] {get; set;}
}

struct S1 : I1
{
    public int F1;

    public int this[int i]
    {
        get
        {
            System.Console.Write(F1);
            return 0;
        }
        set
        {
            System.Console.Write(F1);
        }
    }
}

class Program
{
    static async Task Main()
    {
        await Test3<S1>();
    }

    static T GetT<T>() where T : I1 => (T)(object)new S1 { F1 = 123 };

    static async Task Test3<T>() where T : I1
    {
       GetT<T>()[0] += await Get1Async();
    }

    static async Task<int> Get1Async()
    {
        await Task.Yield();
        return 1;
    }
}
```

Observed: 
```
// /Program.cs(37,8): error CS8178: A reference returned by a call to 'Program.GetT<T>()' cannot be preserved across 'await' or 'yield' boundary.
//        GetT<T>()[0] += await Get1Async();
Diagnostic(ErrorCode.ERR_RefReturningCallAndAwait, "GetT<T>()").WithArguments("Program.GetT<T>()").WithLocation(37, 8)
```

```Program.GetT<T>``` doesn't return a reference. Async rewriter should have no problem capturing the value.

**Comment by bernd5** [comment]
There seems to be to much hoisting. In this case a temporary ref local should be hoisted which was generated by the SpillSequenceSpiller.

Looks similar to: #79416

**Comment by bernd5** [comment]
In my analysis I have observered that there is no issue in the hoisting analysis. The issue seems to be caused by this line:
https://github.com/dotnet/roslyn/blob/62daf28c872ce436a19e790586bd7597eae3e827/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs#L686

This causes that the result of a non-ref method call is captured in a ref local (which does not work - or is at least strange). I observerd such strange codegen earlier - but for non async code such things are later on fixed in the pipeline.

So I would say that a rvalue should be captured by value - regardless if it is a ref or value type.

To improve the codegen we should mark the temporary with `IsKnownToReferToTempIfReferenceType` - otherwise the SpillsequenceSpiller is executing this if block: https://github.com/dotnet/roslyn/blob/62daf28c872ce436a19e790586bd7597eae3e827/src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs#L1038

**Comment by AlekseyTs** [comment]
> This causes that the result of a non-ref method call is captured in a ref local (which does not work - or is at least strange).

In principle, there is nothing wrong or strange in doing this. Non-async scenario works properly, in situations like this codegen simply   captures value in a temp and then get an address of a temp. 

> In my analysis I have observered that there is no issue in the hoisting analysis. 

The error produced "A reference returned by a call to 'Program.GetT<T>()' cannot be preserved across 'await' or 'yield' boundary." is simply wrong, the method doesn't return a reference, the async rewrite shouldn't treat it as one. It should capture the result in a temp as codegen does. I assume the problem is in ```SpillsequenceSpiller```, or the layer that reports the error. 

Since the error is method call specific, I would change the scenario to a different rvalue (not a method call) and check what happens then. If that scenario works, I would see what compiler does different for it (by comparison to a method call).

**Comment by AlekseyTs** [comment]
> To improve the codegen we should mark the temporary with `IsKnownToReferToTempIfReferenceType` - otherwise the SpillsequenceSpiller is executing this if block ...

The purpose of ```IsKnownToReferToTempIfReferenceType``` is to enable some optimizations, not to enforce correctness. It should always be safe to set it to `false` and still have correct behavior of the pipeline and correct code generated at the end. Therefore, its value is definitely not the root cause of the problem here.

**Comment by bernd5** [comment]
> The purpose of IsKnownToReferToTempIfReferenceType is to enable some optimizations, not to enforce correctness. It should always be safe to set it to false and still have correct behavior of the pipeline and correct code generated at the end. Therefore, its value is definitely not the root cause of the problem here.

Of course - without, it would generate unnecessary code - but it does not crash...

**Comment by bernd5** [comment]
The issue can be simplified to:

It produces the same error - because of `rewrittenReceiver.Type.IsValueTyp` in https://github.com/dotnet/roslyn/blob/62daf28c872ce436a19e790586bd7597eae3e827/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs#L686

```csharp
using System.Threading.Tasks;

struct S1
{
    public int this[int i]
    {
        get => 0;
        set => System.Console.Write(value);
    }
}

class Program
{
    static async Task Main()
    {
        GetS1()[0] += await Get1Async();
    }

    static S1 GetS1() => new S1();

    static async Task<int> Get1Async()
    {
        await Task.Yield();
        return 1;
    }
}
```

**Comment by bernd5** [comment]
> In principle, there is nothing wrong or strange in doing this. Non-async scenario works properly, in situations like this codegen simply captures value in a temp and then get an address of a temp.

The strange thing is that you can't initialize a ref local with a rvalue. To do so you need to store that rvalue somewhere, e.g. in an additional local and get a ref to that storage.

To actually compile it you could change the code:
```csharp
refKind = rewrittenReceiver.Type.IsValueType || rewrittenReceiver.Type.Kind == SymbolKind.TypeParameter ? RefKind.Ref : RefKind.None; 
```
to:

```csharp
switch (rewrittenReceiver)
{
    case BoundCall call when call.Method.RefKind == RefKind.None:
        refKind = RefKind.None;
        break;
    default:
        refKind = rewrittenReceiver.Type.IsValueType || rewrittenReceiver.Type.Kind == SymbolKind.TypeParameter ? RefKind.Ref : RefKind.None;
        break;
}
```

**Comment by bernd5** [comment]
The actual diagnostic is emitted here: https://github.com/dotnet/roslyn/blob/e55e7c1019269a18f40661a4e96d836367797871/src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs#L650

**Comment by AlekseyTs** [comment]
> To actually compile it you could change the code: ...

To be clear. I think the problem is not in the LocalRewriter as long as non async version of the scenario works.

**Comment by bernd5** [comment]
Sure, even if the lowering produces over complicated code, it should not crash.

If you look at the [decompiled non async code](https://lab.razor.fyi/#pVE9SwNBEEVJoVuJv2DKO4QjazovSSNEIghCFgRDwMtmcllN9mRnYpBwrY3_xX9kbWPhT5C7W_OlnVvNzux7895b8VkT4tplqUtmkabj99qcjE2h90yMs0hNHCYjY9NIJfRAsRDGMrpxohG6EsRSAAAYy8ATQ_2iMANYpsgxEHKci1wIYjfXDD0JZ9CVHvM4H06NLqEdGYvd3gZdOapAxUmRV_W6Wxyv-TyzlE0xunGGMejIMN565ZDnzkJ93c1XFf2Du2IpDetpQgQ-Ve-XOGGj4SkzI7hKjA3CHV8KiRvNnmwHnjSvQvFABRfIqqnaQQiLCToEVcYJrTYEKgyy4T1qDi0uiqCX0JHQAnnagNyHuymg2vWLa1vRal-_PoCTVnGXf2sr_qma7njyWcsfUH65r-n28ODr7fXjZXxUu9v7Bg) you get:

```csharp
private static void Test3<T>() where T : I1
{
	T t = GetT<T>();
	ref T reference = ref t;
	T val = default(T);
	if (val == null)
	{
		val = reference;
		reference = ref val;
	}
	reference[0] = t[0] + Get1();
}
```

Why can't it be simplified to:
```csharp
private static void Test3<T>() where T : I1
{
	T t = GetT<T>();
	t[0] = t[0] + Get1();
}
```

**Comment by AlekseyTs** [comment]
It can be, however, depending on how that is achieved, that might not fix the root cause of the problem reported in this issue, but might simply mask/hide it. In context of this issue, I think we would prefer fixing the root cause of the problem. Simply put, the pipeline should be able to handle this shape of the tree.
